stdlib/public/core/Array.swift

```diff
@@ -1264,7 +1263,6 @@ extension Array: RangeReplaceableCollection {
         _internalInvariant(remainder.next() == nil)
```

Comment:
Just for my understanding: `if T.self == U.self` in general deals with metatypes and might not work in other places in Embedded Swift (especially when building with -Onone)... correct? I think that we talked about the option of adding a builtin (that tests for equality) that would just need thin metatypes... would that be a robust solution or would that still be problematic?

Reply:
We already have this builtin (`is_same_metatype`). I just realized that here the builtin is used. So the cast simplification is not really required for this change in the array code.
Anyway, this cast simplification enables casts of generic types in generic functions, because such cast can be constant folded once the generic function is specialized.